### PR TITLE
Draft proposal for join command

### PR DESCRIPTION
**Pull request by moelsayed**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: docs/proposals/20200329-join.md [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+* Parse options and node configuration, create state
+  * Apply defaults on the node
+
+### Join
+* Continuing after the initial steps
+* Run the following tasks on the new node:
+  * `installPrerequisitesOnNode`
+  * `generateKubeadmOnNode`
+  * `joinStaticWorkerInternal`
+* Store node configuration in a ConfigMap/machine object upon successful join.
```
Comment (moelsayed) [inline docs/proposals/20200329-join.md]: Not very happy about this. State shouldn't be stored on the cluster. So, open to ideas!

File: docs/proposals/20200329-join.md [hunk 2]
```diff
@@ -0,0 +1,48 @@
+# KubeOne Join
```
Comment (xmudrii) [inline docs/proposals/20200329-join.md]: I'd rename this proposal to something like `KubeOne Static Workers` or `Provisioning Worker Nodes Using KubeOne` to make it clear what this proposal is about.

File: docs/proposals/20200329-join.md [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+PR [#834](https://github.com/kubermatic/kubeone/pull/834) adds support for configuring static worker nodes to be deployed during initial cluster provisioning.
+
+The `join` command should provide allow users to easily add additional worker nodes to an existing KubeOne cluster.
+
+Currently, this is only doable by applying several manual steps on each worker node to deploy the required packages/files and execute `kubeadm join`.
+
+This new subcommand should allow users to add new static worker nodes to the clusters with a single command, and include these nodes in upgrade process as well.
+
+## Goals
+* Implement the `join` subcommand.
```
Comment (xmudrii) [inline docs/proposals/20200329-join.md]: To my understanding, the plan was that we don't want to have the `join` command, but instead to extend the `install` command (like in #834). Do we actually want the `join` command? Has the plan changed or I misunderstood something?

File: docs/proposals/20200329-join.md [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+  * Add the ability to join a static worker node to an existing KubeOne cluster.
+  * Support upgrade process for the added static nodes. These nodes will _not_ be in the cluster configuration file and this should be handled correctly.
```
Comment (xmudrii) [inline docs/proposals/20200329-join.md]: What is the reason for not having those nodes in the cluster configuration file? If we decide to take on this approach, this should be documented somewhere.

File: docs/proposals/20200329-join.md [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+## Non-goals
+* Support adding control plane nodes.
```
Comment (xmudrii) [inline docs/proposals/20200329-join.md]: I don't understand this non-goal. We already support adding control plane nodes, but we don't support replacing and rotating. Is this what you meant with this point?

File: docs/proposals/20200329-join.md [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+* Support node removal/cleanup. This can be done manually by the user.
```
Comment (xmudrii) [inline docs/proposals/20200329-join.md]: IMO, we should extend the `kubeone reset` command to allow this. Once you unprovision the control plane, you can't use worker nodes until you don't clean them up.

File: docs/proposals/20200329-join.md [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+
+## Implementation
+* Require standard subcommand arguments
+* Require new node configurations
+  * minimal required CLI options to join the nodes
+  * `HostConfig` YAML manifest for the node
```
Comment (xmudrii) [inline docs/proposals/20200329-join.md]: I think we should extend the core KubeOneCluster API like in #843. I think the best approach is to have a single source of truth (`config.yaml`) and store all cluster configuration there.

File: docs/proposals/20200329-join.md (hunk 1 above)
Comment (xmudrii) [reply docs/proposals/20200329-join.md]: I'm also not happy about this at all. This feature should work like we handle control planes. Have a YAML manifest where you would store the config needed to provision those nodes. Then, it's the operator's responsibility to store and keep that file.

**Comment by moelsayed** [comment]
https://github.com/kubermatic/kubeone/issues/816

**Comment by moelsayed** [comment]
As discussed offline, we will go with Static Workers #834 instead of this.

**Pull request closed without merging**